build: add `-W*-whitespace`

fanquake opened the pull request:
GCC 15 now has options to turn leading & trailing whitespace into compile failures: https://gcc.gnu.org/gcc-15/changes.html#c-family. Fix the few cases of leading tabs, and trailing whitespace, and then enable `-Wleading-whitespace` and `-Wtrailing-whitespace`.

We currently get PRs that are opened with various whitespace, i.e #33822, so turning that into compile-time failure where possible, seems useful, to avoid a CI roundtrip.

fanquake committed:
build: add -Wleading-whitespace=spaces

This is available in GCC 15. See
https://gcc.gnu.org/gcc-15/changes.html.

fanquake committed:
scripted-diff: change whitespace to spaces in univalue

-BEGIN VERIFY SCRIPT-
sed -i 's/^\t/    /' src/univalue/include/univalue_escapes.h
-END VERIFY SCRIPT-

Co-authored-by: Anthony Towns <aj@erisian.com.au>

fanquake committed:
scripted-diff: remove whitespace in sha256_sse4.cpp

-BEGIN VERIFY SCRIPT-
expand -t4 src/crypto/sha256_sse4.cpp | sed 's/  *$//' | sponge src/crypto/sha256_sse4.cpp
-END VERIFY SCRIPT-

Co-authored-by: Anthony Towns <aj@erisian.com.au>

hebasto commented:
> GCC 15 now has options to turn leading & trailing whitespace into compile failures: https://gcc.gnu.org/gcc-15/changes.html#c-family. Fix the few cases of leading tabs, and trailing whitespace, and then enable `-Wleading-whitespace`. Unfortunately, CMake/Qt will generate code that contains trailing whitespace:
> 
> ```shell
> [ 98%] Building CXX object src/qt/CMakeFiles/bitcoinqt.dir/bitcoinqt_autogen/EJRQKI7XPS/qrc_bitcoin_locale.cpp.o
> /root/bitcoin/build/src/qt/bitcoinqt_autogen/EWIEGA46WW/qrc_bitcoin.cpp:5603:2: warning: trailing whitespace [-Wtrailing-whitespace=]
>  5603 | 
> /root/bitcoin/build/src/qt/bitcoinqt_autogen/EWIEGA46WW/qrc_bitcoin.cpp:8254:2: warning: trailing whitespace [-Wtrailing-whitespace=]
>  8254 | 
> /root/bitcoin/build/src/qt/bitcoinqt_autogen/EWIEGA46WW/qrc_bitcoin.cpp:9840:2: warning: trailing whitespace [-Wtrailing-whitespace=]
> ```
> 
> So `-Wtrailing-whitespace` could be enabled later, once Qts tooling is fixed.

CMake 3.27 and newer handle this. The [`CMP0151`](https://cmake.org/cmake/help/latest/policy/CMP0151.html) policy must be set to `NEW`.

hebasto reviewed (approved):
ACK 5e2620443d13b58efb38310263d67e2b823d6a1e.

The first two commits were reviewed using `git diff --ignore-space-change`. Commit f2385191520f35fdef07743e5a12feaa87085dbe additionally removes an empty line, which is OK.

> > So `-Wtrailing-whitespace` could be enabled later, once Qts tooling is fixed.
> 
> CMake 3.27 and newer handle this. The [`CMP0151`](https://cmake.org/cmake/help/latest/policy/CMP0151.html) policy must be set to `NEW`.

~As it is reasonable to expect CMake ≥3.27 on systems with GCC 15, I believe `-Wtrailing-whitespace` can be reintroduced.~

fanquake committed:
build: add -Wtrailing-whitespace=any

This is available in GCC 15. See
https://gcc.gnu.org/gcc-15/changes.html.

l0rinc commented on src/crypto/sha256_sse4.cpp:
```diff
@@ -1020,15 +1020,15 @@ void Transform(uint32_t* s, const unsigned char* chunk, size_t blocks)
 %macro addm 2
-    add	%2, %1
-    mov	%1, %2
+    add %2, %1
```
nit: given that we're fixing non-trailing tabs as well, the only remaining .cpp/.h file that still contains tabs is https://github.com/bitcoin/bitcoin/blob/75a5c8258ec5309fe506438aa3815608430b53d6/src/util/subprocess.h#L1149, if you edit again, consider including that

fanquake replied:
Sure. https://github.com/arun11299/cpp-subprocess/pull/121.

l0rinc commented on src/crypto/sha256_sse4.cpp:
```diff
@@ -969,27 +969,27 @@ void Transform(uint32_t* s, const unsigned char* chunk, size_t blocks)
-; 
-; All rights reserved. 
-; 
+; Copyright (c) 2012, Intel Corporation
```
nit: there are a few remaining trailing spaces e.g. in https://github.com/bitcoin/bitcoin/blob/07f0a61ef711a2f75ded3d73545bfabdf2a64fef/src/minisketch/src/fields/clmul_common_impl.h#L39, we could add a PR to https://github.com/bitcoin-core/minisketch to get rid of those as well

fanquake replied:
https://github.com/bitcoin-core/minisketch/pull/98.

l0rinc reviewed (approved):
Core review ACK 8c60dc39a353c9d66cbe3a5a0a934eca6b2a287a

fanquake committed:
subprocess: replace tab with space

See: https://github.com/arun11299/cpp-subprocess/pull/121.

l0rinc commented:
ACK b271b342e36000952a389c05b409bab691a4e4d2

fanquake committed:
ci: add moreutils to lint job

Makes the sponge utility available for an upcoming scripted-diff.

hebasto reviewed (commented):
Tested 2e713c22c67b3551f04273428db266503178e7fa on Ubuntu 25.10 with GCC 15.2.0:
```
$ cmake --build build
<snip>
[340/1092] Building CXX object src/qt/CMakeFiles/bitcoinqt.dir/bitcoinqt_autogen/EWIEGA46WW/qrc_bitcoin.cpp.o
src/qt/bitcoinqt_autogen/EWIEGA46WW/qrc_bitcoin.cpp:5603:2: warning: trailing whitespace [-Wtrailing-whitespace=]
 5603 | 
src/qt/bitcoinqt_autogen/EWIEGA46WW/qrc_bitcoin.cpp:8254:2: warning: trailing whitespace [-Wtrailing-whitespace=]
 8254 | 
src/qt/bitcoinqt_autogen/EWIEGA46WW/qrc_bitcoin.cpp:9840:2: warning: trailing whitespace [-Wtrailing-whitespace=]
 9840 | 
src/qt/bitcoinqt_autogen/EWIEGA46WW/qrc_bitcoin.cpp:11671:2: warning: trailing whitespace [-Wtrailing-whitespace=]
11671 | 
src/qt/bitcoinqt_autogen/EWIEGA46WW/qrc_bitcoin.cpp:11880:2: warning: trailing whitespace [-Wtrailing-whitespace=]
11880 | 
src/qt/bitcoinqt_autogen/EWIEGA46WW/qrc_bitcoin.cpp:32981:2: warning: trailing whitespace [-Wtrailing-whitespace=]
32981 | 
src/qt/bitcoinqt_autogen/EWIEGA46WW/qrc_bitcoin.cpp:33585:2: warning: trailing whitespace [-Wtrailing-whitespace=]
33585 | 
src/qt/bitcoinqt_autogen/EWIEGA46WW/qrc_bitcoin.cpp:34942:2: warning: trailing whitespace [-Wtrailing-whitespace=]
34942 | 
src/qt/bitcoinqt_autogen/EWIEGA46WW/qrc_bitcoin.cpp:35370:2: warning: trailing whitespace [-Wtrailing-whitespace=]
35370 | 
[348/1092] Building CXX object src/qt/CMakeFiles/bitcoinqt.dir/bitcoinqt_autogen/EJRQKI7XPS/qrc_bitcoin_locale.cpp.o
src/qt/bitcoinqt_autogen/EJRQKI7XPS/qrc_bitcoin_locale.cpp:82738:2: warning: trailing whitespace [-Wtrailing-whitespace=]
82738 | 
src/qt/bitcoinqt_autogen/EJRQKI7XPS/qrc_bitcoin_locale.cpp:93715:2: warning: trailing whitespace [-Wtrailing-whitespace=]
93715 | 
src/qt/bitcoinqt_autogen/EJRQKI7XPS/qrc_bitcoin_locale.cpp:97857:2: warning: trailing whitespace [-Wtrailing-whitespace=]
97857 | 
src/qt/bitcoinqt_autogen/EJRQKI7XPS/qrc_bitcoin_locale.cpp:204993:2: warning: trailing whitespace [-Wtrailing-whitespace=]
204993 | 
src/qt/bitcoinqt_autogen/EJRQKI7XPS/qrc_bitcoin_locale.cpp:212562:2: warning: trailing whitespace [-Wtrailing-whitespace=]
212562 | 
src/qt/bitcoinqt_autogen/EJRQKI7XPS/qrc_bitcoin_locale.cpp:228433:2: warning: trailing whitespace [-Wtrailing-whitespace=]
228433 | 
src/qt/bitcoinqt_autogen/EJRQKI7XPS/qrc_bitcoin_locale.cpp:228777:2: warning: trailing whitespace [-Wtrailing-whitespace=]
228777 | 
src/qt/bitcoinqt_autogen/EJRQKI7XPS/qrc_bitcoin_locale.cpp:228933:2: warning: trailing whitespace [-Wtrailing-whitespace=]
228933 | 
[419/1092] Building CXX object src/CMakeFiles/minisketch.dir/minisketch/src/fields/clmul_4bytes.cpp.o
In file included from ../src/minisketch/src/fields/clmul_4bytes.cpp:12:
../src/minisketch/src/fields/clmul_common_impl.h:39:66: warning: trailing whitespace [-Wtrailing-whitespace=]
   39 |         __m128i red1 = _mm_clmulepi64_si128(high1, MOD128, 0x00);
[1092/1092] Linking CXX executable bin/fuzz
```

hebasto commented:
> I guess we are now just blocked on Qt / GUI tooling...

From https://github.com/bitcoin/bitcoin/pull/32648#issuecomment-3027824977:
> this should be fixed in Qts tooling.

Even if it’s been fixed upstream, I don’t expect that change to be backported to all Qt versions down to 6.2. Therefore, we still need to suppress warnings in generated files. For example, as follows:
```diff
--- a/src/qt/CMakeLists.txt
+++ b/src/qt/CMakeLists.txt
@@ -50,11 +50,20 @@ endfunction()
 
 set(CMAKE_AUTOMOC ON)
 set(CMAKE_AUTOMOC_MOC_OPTIONS "-p${CMAKE_CURRENT_SOURCE_DIR}")
-set(CMAKE_AUTORCC ON)
 set(CMAKE_AUTOUIC ON)
 set(CMAKE_AUTOUIC_SEARCH_PATHS forms)
 
+set(CMAKE_AUTORCC OFF)
 configure_file(bitcoin_locale.qrc bitcoin_locale.qrc USE_SOURCE_PERMISSIONS COPYONLY)
+qt6_add_resources(BITCOIN_QRC bitcoin.qrc)
+qt6_add_resources(BITCOIN_LOCALE_QRC ${CMAKE_CURRENT_BINARY_DIR}/bitcoin_locale.qrc)
+set_property(
+  SOURCE
+    ${BITCOIN_QRC}
+    ${BITCOIN_LOCALE_QRC}
+  PROPERTY COMPILE_OPTIONS
+    -Wno-trailing-whitespace
+)
 
 # The bitcoinqt sources have to include headers in
 # order to parse them to collect translatable strings.
@@ -119,8 +128,8 @@ add_library(bitcoinqt STATIC EXCLUDE_FROM_ALL
   utilitydialog.h
   $<$<PLATFORM_ID:Windows>:winshutdownmonitor.cpp>
   $<$<PLATFORM_ID:Windows>:winshutdownmonitor.h>
-  bitcoin.qrc
-  ${CMAKE_CURRENT_BINARY_DIR}/bitcoin_locale.qrc
+  ${BITCOIN_QRC}
+  ${BITCOIN_LOCALE_QRC}
 )
 target_compile_definitions(bitcoinqt
   PUBLIC
```

fanquake commented:
> Even if it’s been fixed upstream,

It's not clear to me if it has been fixed or not, can you link to the relevant change / issue? I can cherry-pick the change above in, if you create a commit, and add a comment inline, for when the change can be dropped.

hebasto committed:
cmake: Switch from AUTORCC to `qt6_add_resources`

This change provides finer control over the compilation of RCC-generated
files.

hebasto committed:
cmake: Disable `-Wtrailing-whitespace` warnings for RCC-generated files

See: https://bugreports.qt.io/browse/QTBUG-141858.

hebasto reviewed (approved):
ACK e412d6d438d600267174d1281feae3954c1c1c5b, tested on Ubuntu 25.10 using GCC 15.

fanquake commented on CMakeLists.txt:
```diff
@@ -463,6 +463,9 @@ else()
   try_append_cxx_flags("-Wundef" TARGET warn_interface SKIP_LINK)
+  try_append_cxx_flags("-Wleading-whitespace=spaces" TARGET warn_interface SKIP_LINK)
+  try_append_cxx_flags("-Wtrailing-whitespace=any" TARGET warn_interface SKIP_LINK)
+
```
Dropped.

hebasto reviewed (approved):
re-ACK 40dcbf580d8eb31a067b62bf9676099919b9841e.

l0rinc commented:
> how many devs are using GCC-15

lately I'm also doing most of my benchmarks with 15

maflcko commented:
Not sure how to reproduce locally, but this seems to be failing:

```
/__w/bitcoin-core-nightly/bitcoin-core-nightly/src/clientversion.cpp:33:79: error: trailing whitespace [-Werror=trailing-whitespace=]
   33 | //! git will put "#define GIT_COMMIT_ID ..." on the next line inside archives.
```

https://github.com/maflcko/bitcoin-core-nightly/actions/runs/19297432422/job/55182808314#step:9:403

fanquake commented:
> See https://github.com/bitcoin-core/secp256k1/pull/1813.

Shouldn't this be fixed by not applying these flags to secp256k1 code and/or fixing passing down `-Werror`, or both?

hebasto commented:
> > See [bitcoin-core/secp256k1#1813](https://github.com/bitcoin-core/secp256k1/pull/1813).
> 
> Shouldn't this be fixed by not applying these flags to secp256k1 code and/or fixing passing down `-Werror`, or both?

I doubt that any of these measures would fix the issue because the trailing space is in a header, causing an error when compiling Core's object files.

fanquake commented:
Opened #34442.
